lib/src/main/java/graphql/nadel/engine/instrumentation/NadelInstrumentationTimer.kt

```diff
@@ -0,0 +1,100 @@
+package graphql.nadel.enginekt.instrumentation
+
+import graphql.execution.instrumentation.InstrumentationState
+import graphql.nadel.instrumentation.NadelInstrumentation
+import graphql.nadel.instrumentation.parameters.NadelInstrumentationTimingParameters
+import graphql.nadel.instrumentation.parameters.NadelInstrumentationTimingParameters.Step
+import java.time.Duration
+
+internal class NadelInstrumentationTimer(
+    private val instrumentation: NadelInstrumentation,
+    private val userContext: Any?,
+    private val instrumentationState: InstrumentationState?,
+) {
+    inline fun <T> time(
+        step: Step,
+        function: () -> T,
+    ): T {
+        val start = System.nanoTime()
+
+        try {
+            return function()
+                .also {
+                    val duration = getDurationSince(start)
+                    instrumentation.onStepTimed(newParameters(step, duration))
+                }
+        } catch (e: Throwable) {
+            try {
+                val duration = getDurationSince(start)
+                instrumentation.onStepTimed(newParameters(step, duration, exception = e))
+            } catch (e2: Throwable) {
+                e2.addSuppressed(e)
+                throw e2
+            }
+
+            throw e
+        }
+    }
+
+    fun batch(): BatchTimer {
+        return BatchTimer(timer = this)
+    }
+
+    inline fun <T> batch(function: (BatchTimer) -> T): T {
```

Comment:
Ok I added this so that we don't get a bunch of <1ms times. So the way this works is you can do this

```kotlin
timer.batch { batchTimer ->
  batchTimer.time(step = X) {
    delay(1024)
  }
  batchTimer.time(step = X) {
    delay(1024)
  }
}
```

And at the end X is emitted with 2048 i.e. the two `time` calls are combined.

We use this to collate the time taken for each transformation even though it might be executed multiple times for different fields.